### PR TITLE
SSH task exit code added to XCOM as 'ssh_exit' key

### PR DESCRIPTION
**Pull request by kolfild26**
Based on https://github.com/apache/airflow/discussions/23788

**Motivation.** Provide the ability to analyze the SSH tasks exit code in order to build a dag logic based on these codes. Now, only `echo` output goes to XCOM's return_value key. And in case of failure nothing being passed to return_value. So in a dag we don't know what specific code is returned.

**What's done:**
In [SSHOperator](https://github.com/apache/airflow/blob/main/airflow/providers/ssh/operators/ssh.py#L36), add XCOM key **ssh_exit**.
Task succeeded - ssh_exit set to 0. `echo` stuff still passed to return_value.
Task failed - ssh_exit stores the code which was retrived from the ssh session.

```
SSHOperator(
        command="pwd",
        dag=dag)

SSHOperator(
        command="pwd && exit 123",
        dag=dag)
```

изображение

**P.S.**  I could imagine the alternative possible scenario - not to introduce a new XCOM key and just replace the return_value content, i.e. skip all other text output and just pass the ssh exit code there.
But my first approach does not reduce the functionality and the second one does. So I prefered the first.

### CODE REVIEW

File: airflow/providers/ssh/operators/ssh.py [hunk 1]
```diff
@@ -151,16 +151,18 @@ def exec_ssh_client_command(self, ssh_client: SSHClient, command: str):
             ssh_client, command, timeout=self.timeout, environment=self.environment, get_pty=self.get_pty
         )
 
-    def raise_for_status(self, exit_status: int, stderr: bytes) -> None:
+    def raise_for_status(self, exit_status: int, stderr: bytes, **kwargs) -> None:
+        ti=kwargs['context'].get('task_instance')
+        ti.xcom_push(key='ssh_exit', value=exit_status)
```
Comment (bdsoha) [inline airflow/providers/ssh/operators/ssh.py]:
```suggestion
        ti=kwargs["context"].get("task_instance")
        ti.xcom_push(key="ssh_exit", value=exit_status)
```

File: airflow/providers/ssh/operators/ssh.py [hunk 2]
```diff
@@ -151,16 +151,18 @@ def exec_ssh_client_command(self, ssh_client: SSHClient, command: str):
...
             raise AirflowException(f"SSH operator error: exit status = {exit_status}")
 
-    def run_ssh_client_command(self, ssh_client: SSHClient, command: str) -> bytes:
+    def run_ssh_client_command(self, ssh_client: SSHClient, command: str, **kwargs) -> bytes:
         assert self.ssh_hook
         exit_status, agg_stdout, agg_stderr = self.ssh_hook.exec_ssh_client_command(
             ssh_client, command, timeout=self.timeout, environment=self.environment, get_pty=self.get_pty
         )
-        self.raise_for_status(exit_status, agg_stderr)
+        self.raise_for_status(exit_status, agg_stderr, context=kwargs['context'])
```
Comment (bdsoha) [inline airflow/providers/ssh/operators/ssh.py]:
```suggestion
        self.raise_for_status(exit_status, agg_stderr, context=kwargs["context"])
```

**Review by bdsoha — CHANGES_REQUESTED** [review]
In order to approve the PR a check was added to *Static Checks* requiring the use of double-quotes.

**Comment by kolfild26** [comment]
@bdsoha thanks, applied

File: airflow/providers/ssh/operators/ssh.py [hunk 3]
```diff
@@ -151,16 +151,18 @@ def exec_ssh_client_command(self, ssh_client: SSHClient, command: str):
             ssh_client, command, timeout=self.timeout, environment=self.environment, get_pty=self.get_pty
         )
 
-    def raise_for_status(self, exit_status: int, stderr: bytes) -> None:
+    def raise_for_status(self, exit_status: int, stderr: bytes, **kwargs) -> None:
+        ti=kwargs["context"].get("task_instance")
+        ti.xcom_push(key="ssh_exit", value=exit_status)
```
Comment (eladkal) [inline airflow/providers/ssh/operators/ssh.py]: I think we should verify user wish to push to xcom.
users ask for this specifically with `do_xcom_push` parameter

**Review by eladkal — COMMENTED** [review]
Please add unit test to cover this change

File: airflow/providers/ssh/operators/ssh.py (hunk 3 above)
Comment (bdsoha) [reply airflow/providers/ssh/operators/ssh.py]: @eladkal It is not optional on other operators:

https://github.com/apache/airflow/blob/b29ca4e77d4d80fb1f4d6d4b497a3a14979dd244/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py#L444-L446

**Review by potiuk — CHANGES_REQUESTED** [review]
Yep. It should have unit tests added.

**Comment by kolfild26** [comment]
in progress. ut will be added.

File: airflow/providers/ssh/operators/ssh.py [hunk 4]
```diff
@@ -134,16 +134,18 @@ def exec_ssh_client_command(self, ssh_client: SSHClient, command: str):
             ssh_client, command, timeout=self.cmd_timeout, environment=self.environment, get_pty=self.get_pty
         )
 
-    def raise_for_status(self, exit_status: int, stderr: bytes) -> None:
+    def raise_for_status(self, exit_status: int, stderr: bytes, **kwargs) -> None:
+        ti=kwargs["context"].get("task_instance")
```
Comment (bdsoha) [inline airflow/providers/ssh/operators/ssh.py]:
```suggestion
        ti = kwargs["context"].get("task_instance")
```

**Comment by kolfild26** [comment]
Hi,
looks like, it's done now. Please review changes.

1. Fixed errors
`pre-commit run run-mypy` and `breeze testing tests --test-type "Providers[ssh]"` were successful.
2. Description added
3. [test_push_ssh_exit_to_xcom](https://github.com/apache/airflow/blob/6627c9b0c31b3ce3024abca004a995b55a63ba80/tests/providers/ssh/operators/test_ssh.py#L238) for the newly introduced functionality added

**Comment by kolfild26** [comment]
Also some notes about the test which I added.
There were no any tests which uses `conext` inside themselves.
I, in turn, needed `context` to pull/push to/from xcom. So, I searched all from [the tests](https://github.com/apache/airflow/tree/main/tests) and found that [test_kubernetes_pod.py](https://github.com/apache/airflow/blob/main/tests/providers/cncf/kubernetes/operators/test_kubernetes_pod.py) has [create_context method](https://github.com/apache/airflow/blob/main/tests/providers/cncf/kubernetes/operators/test_kubernetes_pod.py#L57). It turned out it fully fitts our need. So I copied it to `test_ssh.py` and used in [test_push_ssh_exit_to_xcom](https://github.com/apache/airflow/blob/6627c9b0c31b3ce3024abca004a995b55a63ba80/tests/providers/ssh/operators/test_ssh.py#L238).

**Review by potiuk — APPROVED** [review]
Looks really cool now :). Very much needed. Let's see if our CI agrees.

**Comment by kolfild26** [comment]
@potiuk  Static checks error in the [#59009](https://github.com/apache/airflow/actions/runs/3509076878) run relates to `sensors/base.py` which was not modified in my commits.
Probably that's because last time I pulled from main 7 days ago and  [#27871](https://github.com/apache/airflow/pull/27871) was pushed 3 days ago i.e you ran the version which doesn't contain these changes.
If it's needed I update my PR branch just now.

**Comment by potiuk** [comment]
Yep. Rebasing is alwasy a good idea.

**Comment by kolfild26** [comment]
[Merged branch 'main' into ssh_exit_code](https://github.com/apache/airflow/pull/27370/commits/ea5afa96155c52efff4201d085fdf17caf9fcb76) ✅

**Pull request merged**